pkg/agent/controller/bgp/controller.go

```diff
@@ -0,0 +1,1015 @@
+
+	queue workqueue.RateLimitingInterface
+}
+
+func NewBGPPolicyController(ctx context.Context,
+	nodeInformer coreinformers.NodeInformer,
+	serviceInformer coreinformers.ServiceInformer,
+	egressInformer crdinformersv1b1.EgressInformer,
+	bgpPolicyInformer crdinformersv1a1.BGPPolicyInformer,
+	endpointSliceInformer discoveryinformers.EndpointSliceInformer,
+	k8sClient kubernetes.Interface,
+	bgpPeerPasswordsSecret string,
+	nodeConfig *config.NodeConfig,
+	networkConfig *config.NetworkConfig) (*Controller, error) {
+	c := &Controller{
+		ctx:                       ctx,
+		nodeInformer:              nodeInformer.Informer(),
+		nodeLister:                nodeInformer.Lister(),
+		nodeListerSynced:          nodeInformer.Informer().HasSynced,
+		serviceInformer:           serviceInformer.Informer(),
+		serviceLister:             serviceInformer.Lister(),
+		serviceListerSynced:       serviceInformer.Informer().HasSynced,
+		egressInformer:            egressInformer.Informer(),
+		egressLister:              egressInformer.Lister(),
+		egressListerSynced:        egressInformer.Informer().HasSynced,
+		bgpPolicyInformer:         bgpPolicyInformer.Informer(),
+		bgpPolicyLister:           bgpPolicyInformer.Lister(),
+		bgpPolicyListerSynced:     bgpPolicyInformer.Informer().HasSynced,
+		endpointSliceLister:       endpointSliceInformer.Lister(),
+		endpointSliceListerSynced: endpointSliceInformer.Informer().HasSynced,
+		bgpPolicyBinding:          &nodeToBGPPolicyBinding{alternativeBPs: sets.Set[string]{}},
+		bgpPolicyStates:           make(map[string]*bgpPolicyState),
+		k8sClient:                 k8sClient,
+		bgpPeerPasswordsSecret:    bgpPeerPasswordsSecret,
+		bgpPeerPasswords:          make(map[string]string),
+		nodeName:                  nodeConfig.Name,
+		enabledIPv4:               networkConfig.IPv4Enabled,
+		enabledIPv6:               networkConfig.IPv6Enabled,
+		podIPv4CIDR:               nodeConfig.PodIPv4CIDR.String(),
+		podIPv6CIDR:               nodeConfig.PodIPv6CIDR.String(),
+		nodeIPv4Addr:              nodeConfig.NodeIPv4Addr.IP.String(),
+		egressEnabled:             features.DefaultFeatureGate.Enabled(features.Egress),
+		queue:                     workqueue.NewNamedRateLimitingQueue(workqueue.NewItemExponentialFailureRateLimiter(minRetryDelay, maxRetryDelay), "bgpPolicyGroup"),
+	}
+	c.bgpPolicyInformer.AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    c.addBGPPolicy,
+			UpdateFunc: c.updateBGPPolicy,
+			DeleteFunc: c.deleteBGPPolicy,
+		},
+		resyncPeriod,
+	)
+	c.serviceInformer.AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    c.addService,
+			UpdateFunc: c.updateService,
+			DeleteFunc: c.deleteService,
+		},
+		resyncPeriod,
+	)
+	c.egressInformer.AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    c.addEgress,
+			UpdateFunc: c.updateEgress,
+			DeleteFunc: c.deleteEgress,
+		},
+		resyncPeriod,
+	)
```

Comment:
I feel we can skip this if egressEnabled is false?

Reply:
You are right. Thanks for pointing out this.